File: medium/1807-evaluate-the-bracket-pairs-of-a-string/solution.py
class Solution(object):
    def evaluate(self, s, knowledge):
        """
        :type s: str
        :type knowledge: List[List[str]]
        :rtype: str
        """

        # Convert knowledge into dictionary
        data = {}

        for key, value in knowledge:
            data[key] = value

        result = []
        i = 0

        while i < len(s):
            if s[i] == '(':
                j = i + 1

                while s[j] != ')':
                    j += 1

                key = s[i + 1:j]

                if key in data:
                    result.append(data[key])
                else:
                    result.append('?')

                i = j + 1

            else:
                result.append(s[i])
                i += 1

        return ''.join(result)

File: medium/1807-evaluate-the-bracket-pairs-of-a-string/solution.dart
class Solution {
  String evaluate(String s, List<List<String>> knowledge) {
    Map<String, String> map = {};

    for (final pair in knowledge) {
      map[pair[0]] = pair[1];
    }

    StringBuffer result = StringBuffer();

    int i = 0;

    while (i < s.length) {
      if (s[i] == '(') {
        int j = i + 1;

        while (s[j] != ')') {
          j++;
        }

        String key = s.substring(i + 1, j);

        result.write(map[key] ?? '?');

        i = j + 1;
      } else {
        result.write(s[i]);
        i++;
      }
    }

    return result.toString();
  }
}